docs/iris/src/developers_guide/dask_interface.rst

```diff
@@ -0,0 +1,35 @@
+Iris Dask Interface
+*******************
+
+Iris uses `dask <http://dask.pydata.org>`_ to manage lazy data interfaces and processing graphs.
+The key principles that define this interface are:
+
+* A call to :attr:`cube.data` will always load all of the data.
+
+* Once this has happened:
+
+  * :attr:`cube.data` is a mutable NumPy masked array or ``ndarray``, and
+  * ``cube._numpy_array`` is a private NumPy masked array, accessible via :attr:`cube.data`, which may strip off the mask and return a reference to the bare ``ndarray``.
+
+* You can use :attr:`cube.data` to set the data. This accepts:
+
+  * a NumPy array (including masked array), which is assigned to ``cube._numpy_array``, or
+  * a dask array, which is assigned to ``cube._dask_array``, while ``cube._numpy_array`` is set to None.
+
+* ``cube._dask_array`` may be None, otherwise it is expected to be a dask array:
+
+  * this may wrap a proxy to a file collection, or
+  * this may wrap the NumPy array in ``cube._numpy_array``.
+
+* All dask arrays wrap array-like objects where missing data are represented by ``nan`` values:
```

Comment:
Have we lost the means to differentiate between 'nan' values and 'masked' values in that case?
Looks like this is the case:

iris master:
```Python
>>> iris.load_cube('nan_mask_tmp.nc').data
masked_array(data = [1.0 -- nan],
             mask = [False  True False],
       fill_value = 1e+20)
```

This branch:
```Python
>>> iris.load_cube('nan_mask_tmp.nc').data
masked_array(data = [1.0 -- --],
             mask = [False  True  True],
       fill_value = 1e+20)
```

Differentiating between masked values and nan values can be important.  An example:  Regridding a field with masked data to a target with a different coordinate system, where extrapolation is set to 'nan' and takes place due a mismatch between the source and target domains (i.e. not 100% overlap).
Though this behaviour I suspect has not changed, at the point of saving this data to disk and loading it back in, we have lost this information.  We cache data to disk that depends on knowing the difference between a masked value and a 'nan' status.

Reply:
Captured in #2609